---
title: "Coalescence experiment"
author: "Jia Xiu (xibeihenai@gmail.com)"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Coalescence experiment - Microbiome - 16S

### Initiate libraries
```{r load_packages_01, message=FALSE}

rm(list=ls())

library(tidyverse)
library(vegan)
```

### Load raw feature table, taxonomy and metadata file
Load feature table
```{r}

# feature table
com <- read.csv("feature-table.csv", header = 1, row.names = 1, sep = ",")
com$S35[1:10]
dim(com)
class(com)
range(rowSums(com[, 1:3]))

df <- com[apply(com, 1, function(x) !all(x==0)),]
dim(df)

# write.csv(df, "feature_table_nonzeros.csv", row.names = TRUE)
com <- read.csv("feature_table_nonzeros.csv", header = 1,  sep = ",")
com <- com %>% 
  column_to_rownames("X") 
# %>% 
#   rownames_to_column(var = "OTU_ID")
dim(com)
com[1:3, 1:2]

# check the range of total reads per sample
range(colSums(com))
cat("the lowest number of reads", head(sort(rowSums(com))))
head(sort(rowSums(com), decreasing = TRUE))
head(sort(colSums(com)))

```

load metadata file
```{r}
metadata <- read.csv("metadata_16S_04-06-2024.tsv", header = 1, sep = "\t")
head(metadata)

```

load and clean taxonomy table
```{r}

# load taxonomy table. You can also use the the origional taxonomy file. I manually reformatted it a little bit
taxa <- read.table("taxonomy-gg2.tsv", header = 1, sep = "\t")
head(taxa)

# clean the taxonomy table (only keep features/ASVs showed in the feature table)
taxa <- taxa %>%
  filter(OTU_ID %in% row.names(com))
head(taxa)
dim(taxa)

# I didn't find any chloroplast or mitochondria in my samples
# k__Bacteria;p__Cyanobacteria;c__Chloroplast;o__Streptophyta;f__;g__;s__ and
# k__Bacteria;p__Proteobacteria;c__Alphaproteobacteria;o__Rickettsiales;f__mitochondria;g__Sarcandra;s__grandifolia

# filter(grepl('c__Chloroplast', taxonomy))

```



###  rarefaction & rarefy table
```{r}
# plot total reads per sample in a histogram
com <- t(com)
total_reads <- rowSums(com)
df <- as.data.frame(total_reads) %>% 
  rownames_to_column(var = "sample.id") %>% 
  left_join(metadata, by = "sample.id") %>% 
  column_to_rownames(var = "sample.id") %>% 
  arrange(total_reads) 

head(df)
table(df$watertype)

(p <- ggplot(df, aes(x = total_reads)) + 
    geom_histogram(bins = 100, position="identity", color="darkgray", fill="white") +
    geom_density(alpha = .2) +
    theme_classic())

(p <- ggplot(df, aes(x = total_reads, fill = watertype)) + 
    facet_wrap(. ~ catagery) +
    geom_histogram(bins = 100, position="identity", alpha = 0.5) +
    geom_density(alpha = .5) +
    geom_vline(aes(xintercept = 2000), color = "darkred") + 
    scale_color_manual(values=c("#999999", "#E69F00", "#56B4E9", "#009E73")) +
    scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9", "#009E73")) +
    theme_classic())

# Set cutoff by youself.
cutoff <- 5000

set.seed(123)
rare <- rrarefy(com, cutoff)
rare <- rare[, colSums(rare)!=0]
rare <- t(rare)

# after rarefying, we lost 101 features, which might have low fequency and abundance (I need to further verify this)
dim(rare)

# check if rarefy work
range(rowSums(rare))
rare[1:2, 1:2]

# save the feature table for later use
# write.csv(rare, "feature_table_rarefied_5000.csv")

```


2. rarefaction curves
```{r}
# prepare a rarefaction dataset
rarecurve_data <- rarecurve(com, step = 100) 

# extract metrix from the output above (following Pat Schloss's Youtube tutorial)
df <- map_dfr(rarecurve_data, bind_rows) %>% 
  bind_cols(Group = row.names(com), .) %>% 
  pivot_longer(-Group) %>% 
  drop_na() %>% 
  mutate(n_seqs = as.numeric(str_replace(name, "N", ""))) %>% 
  select(-name) %>% 
  rename(sample.id = Group) %>% 
  left_join(metadata, by = "sample.id") 

head(df)
dim(df)

( p <- ggplot(df, aes(x = n_seqs, y = value, group = factor(name))) + 
    geom_line(color = "darkgray", alpha = 0.6) +
    geom_vline(xintercept = cutoff, color = "darkred") +
    facet_grid(. ~ catagery) +
    scale_x_continuous(expand = c(0, 0), limits = c(0, 20000)) +
    scale_y_continuous(expand = c(0, 0), limits = c(0, 3000)) +
    labs(x = "Sampling depth", y = "Reads", title = "") + 
    theme_bw() +
    theme(text = element_text(size = 8),
          legend.background = element_rect(),
          legend.title = element_text(face = "bold"),
          strip.background = element_blank(),
          strip.text = element_text(face = "bold", size = 8),
          axis.ticks = element_line(colour = "grey70", size = 0.2),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)))

# ggsave("rarefaction_curves.pdf", width = 17, height = 6, units = "cm", p, scale = 1.5)
```


Clean metadata not necessary
```{r}

# clean metadata (only keep samples showed in the feature table)
# metadata <- metadata %>%
# filter(sample.id %in% colnames(rare))
# dim(metadata)
# head(metadata)

# save metadata for later use
# write.csv(metadata, "metadata_PMB-FT-NL.csv", row.names=FALSE)
```


### Session Info
```{r}
sessionInfo()
```